sno/pk_generation.py

```diff
@@ -406,3 +452,45 @@ def feature_blobs(self):
             pk = self.decode_path_to_1pk(blob.name)
             if self.pk_filter(pk):
                 yield blob
+
+
+class FeaturePlusHashes:
+    _MORE_THAN_ONE_MATCH = object()
+
+    def __init__(self, feature, schema):
+        self.feature = feature
+        self.schema = schema
+        self.match = None
+
+    def __hash__(self):
+        return id(self)
+
+    def __eq__(self, other):
+        return other is self
+
+    def all_related_hashes(self):
+        value_hashes = []
+        feature_hash = 0
+        # Generate the overall hash of the feature, but in separate pieces.
+        for col in self.schema.non_pk_columns:
+            value = self.feature[col.name]
+            value_hash = hash(msg_pack([col.id, value]))
+            feature_hash ^= value_hash
+            value_hashes.append(value_hash)
+
+        # Add the feature at its overall hash.
+        yield feature_hash
+        # Remove each piece one at a time to make a hash for each variation of the feature missing exactly one field.
+        for value_hash in value_hashes:
+            yield feature_hash ^ value_hash
+
+    def try_add_match(self, other):
+        if self.match is None:
+            self.match = other
+        elif self.match == other:
+            pass
+        else:
+            self.match = self._MORE_THAN_ONE_MATCH
+
+    def is_uniquely_matched(self):
+        return self.match is not None and self.match.match is self
```

Comment:
if `self.match == self. _MORE_THAN_ONE_MATCH` then `self.match.match` will throw an `AttributeError` here?

Reply:
I think I hit this (it's very uncommon though)
```
   [14646] Failed to execute script sno_cli
   Traceback (most recent call last):
     File "sno_cli.py", line 4, in <module>
     File "site-packages/click/core.py", line 829, in __call__
     File "site-packages/click/core.py", line 782, in main
     File "site-packages/sno-0.7.2.dev0+ci.1851.git52675e9d-py3.7.egg/sno/cli.py", line 138, in invoke
     File "site-packages/click/core.py", line 1259, in invoke
     File "site-packages/click/core.py", line 1066, in invoke
     File "site-packages/click/core.py", line 610, in invoke
     File "site-packages/click/decorators.py", line 21, in new_func
     File "site-packages/sno-0.7.2.dev0+ci.1851.git52675e9d-py3.7.egg/sno/init.py", line 329, in import_table
     File "site-packages/sno-0.7.2.dev0+ci.1851.git52675e9d-py3.7.egg/sno/fast_import.py", line 208, in fast_import_tables
     File "site-packages/sno-0.7.2.dev0+ci.1851.git52675e9d-py3.7.egg/sno/fast_import.py", line 271, in write_blobs_to_stream
     File "site-packages/sno-0.7.2.dev0+ci.1851.git52675e9d-py3.7.egg/sno/dataset2.py", line 317, in import_iter_feature_blobs
     File "site-packages/sno-0.7.2.dev0+ci.1851.git52675e9d-py3.7.egg/sno/pk_generation.py", line 266, in features
     File "site-packages/sno-0.7.2.dev0+ci.1851.git52675e9d-py3.7.egg/sno/pk_generation.py", line 303, in _match_similar_features_and_remove
     File "site-packages/sno-0.7.2.dev0+ci.1851.git52675e9d-py3.7.egg/sno/pk_generation.py", line 346, in _fast_pop_similar_pairs
     File "site-packages/sno-0.7.2.dev0+ci.1851.git52675e9d-py3.7.egg/sno/pk_generation.py", line 496, in is_uniquely_matched
   AttributeError: 'object' object has no attribute 'match'
   fatal: stream ends early
```

Reply:
Oops! Fixed it